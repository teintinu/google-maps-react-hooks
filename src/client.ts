
const API_URL = 'https://maps.googleapis.com/maps/api/js'
const CALLBACK_NAME = '__googleMapsApiOnLoadCallback'

/**
 * Creates a Google Maps client. The client object contains all the API methods.
 */

export interface GoogleMapsApiOptions {
  /** API key (required, unless clientID and clientSecret provided). */
  key: string
  /** Maps API for Work client ID. */
  client?: string
  /** Maps API for Work channel. */
  channel?: string
  /** Default language for all queries. */
  language?: Language
  /**
   * The region code, specified as a ccTLD (country code top-level domain) two-character value.
   * Most ccTLD codes are identical to ISO 3166-1 codes, with some exceptions.
   * This parameter will only influence, not fully restrict, search results.
   * If more relevant results exist outside of the specified region, they may be included.
   * When this parameter is used, the country name is omitted from the resulting `formatted_address`
   * for results in the specified region.
   */
  region?: Language
  /** load additional Google Maps libraries */
  libraries?: Library[],
}

/**
 * You can load additional libraries by specifying a libraries parameter in the bootstrap request
 *
 * @see https://developers.google.com/maps/documentation/javascript/libraries
 */
export type Library = 'drawing' | 'geometry' | 'places' | 'visualization'

export interface RateOptions {
  /** Controls rate-limiting of requests. Maximum number of requests per period. (Default: 50). */
  limit?: number
  /** Period for rate limit, in milliseconds. (Default: 1000 ms). */
  period?: number
}

export interface RetryOptions {
  /** If a transient server error occurs, how long to wait before retrying the request, in milliseconds. (Default: 500 ms). */
  interval?: number
}

/**
 * By default the API will attempt to load the most appropriate language based on the users location or browser settings.
 * Some APIs allow you to explicitly set a language when you make a request
 *
 * @see https://developers.google.com/maps/faq#languagesupport
 */
export type Language = (
  /** Arabic */
  'ar' |
  /** Belarusian */
  'be' |
  /** Bulgarian */
  'bg' |
  /** Bengali */
  'bn' |
  /** Catalan */
  'ca' |
  /** Czech */
  'cs' |
  /** Danish */
  'da' |
  /** German */
  'de' |
  /** Greek */
  'el' |
  /** English */
  'en' |
  /** English (Australian) */
  'en-Au' |
  /** English (Great Britain) */
  'en-GB' |
  /** Spanish */
  'es' |
  /** Basque */
  'eu' |
  /** Farsi */
  'fa' |
  /** Finnish */
  'fi' |
  /** Filipino */
  'fil' |
  /** French */
  'fr' |
  /** Galician */
  'gl' |
  /** Gujarati */
  'gu' |
  /** Hindi */
  'hi' |
  /** Croatian */
  'hr' |
  /** Hungarian */
  'hu' |
  /** Indonesian */
  'id' |
  /** Italian */
  'it' |
  /** Hebrew */
  'iw' |
  /** Japanese */
  'ja' |
  /** Kazakh */
  'kk' |
  /** Kannada */
  'kn' |
  /** Korean */
  'ko' |
  /** Kyrgyz */
  'ky' |
  /** Lithuanian */
  'lt' |
  /** Latvian */
  'lv' |
  /** Macedonian */
  'mk' |
  /** Malayalam */
  'ml' |
  /** Marathi */
  'mr' |
  /** Burmese */
  'my' |
  /** Dutch */
  'nl' |
  /** Norwegian */
  'no' |
  /** Punjabi */
  'pa' |
  /** Polish */
  'pl' |
  /** Portuguese */
  'pt' |
  /** Portuguese (Brazil) */
  'pt-BR' |
  /** Portuguese (Portugal) */
  'pt-PT' |
  /** Romanian */
  'ro' |
  /** Russian */
  'ru' |
  /** Slovak */
  'sk' |
  /** Slovenian */
  'sl' |
  /** Albanian */
  'sq' |
  /** Serbian */
  'sr' |
  /** Swedish */
  'sv' |
  /** Tamil */
  'ta' |
  /** Telugu */
  'te' |
  /** Thai */
  'th' |
  /** Tagalog */
  'tl' |
  /** Turkish */
  'tr' |
  /** Ukrainian */
  'uk' |
  /** Uzbek */
  'uz' |
  /** Vietnamese */
  'vi' |
  /** Chinese (Simlified) */
  'zh-CN' |
  /** Chinese (Traditional) */
  'zh-TW'
)

let api: undefined | Promise<typeof google.maps>

export async function getGoogleMapsAPI () {
  if (!api) throw new Error('Please call loadGoogleMapsAPI on app initializaion')
  return api
}

export function loadGoogleMapsAPI (options: GoogleMapsApiOptions) {
  if (api) throw new Error('Please call loadGoogleMapsAPI just once')
  const wnd: any = window
  api = new Promise<typeof google.maps>(function (resolve, reject) {
    const tm = setTimeout(function () {
      wnd[CALLBACK_NAME] = () => 0
      reject(new Error('Could not load the Google Maps API'))
    }, 10000)

    wnd[CALLBACK_NAME] = function () {
      clearTimeout(tm)
      resolve(wnd.google.maps)
      delete wnd[CALLBACK_NAME]
    }

    const scriptElement = document.createElement('script')
    const params = [`callback=${CALLBACK_NAME}`]
    Object.entries(options).forEach(([optionName, optionValue]) => {
      if (options.libraries && optionName === 'libraries') params.push(`libraries=${options.libraries.join(',')}`)
      else params.push(optionName + '=' + optionValue)
    })

    scriptElement.src = API_URL + '?' + params.join('&')

    document.body.appendChild(scriptElement)
  })
  return api
}
