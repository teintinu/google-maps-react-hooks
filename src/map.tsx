import React from 'react'
import useGoogleMap, { MapOptions } from './useGoogleMap'

export interface GoogleMapChildElement {
  api: typeof google.maps,
  map: google.maps.Map,
}

export const Map: React.FunctionComponent<MapOptions> = (opts) => {
  const { children, ...options } = opts
  const { api, map, mapRef, loading } = useGoogleMap(options)

  return (
    <div className='map-container'>
      <div ref={mapRef} className='map-ref' />
      {!loading &&
        api ?
        React.Children.map(children, child => {
          return React.cloneElement(child as React.ReactElement, { map, api })
        }) : 'Error loading Google Maps API'}
    </div>
  )
}
