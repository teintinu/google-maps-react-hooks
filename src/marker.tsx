import React from 'react'
import useGoogleMapMarker, { MarkOptions } from './useGoogleMapMarker'

export const Marker: React.FunctionComponent<MarkOptions> = (opts) => {
  useGoogleMapMarker(opts as any)
  return null
}
