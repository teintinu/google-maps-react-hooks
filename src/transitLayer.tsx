import { useEffect, useState } from 'react'
import { GoogleMapChildElement } from './map'

export interface TansitLayerProps {
  enabled: boolean
}

export function TransitLayer (props: TansitLayerProps): React.ReactElement {
  const [transitLayer, setTransitLayer] = useState()
  const { api, map, enabled } = props as any as GoogleMapChildElement & TansitLayerProps
  useEffect(() => {
    setTransitLayer(new api.TransitLayer())
  }, [])

  useEffect(
    () => {
      if (transitLayer) {
        enabled ? transitLayer.setMap(map) : transitLayer.setMap(null)
      }
    },
    [enabled]
  )
  return null as any
}
