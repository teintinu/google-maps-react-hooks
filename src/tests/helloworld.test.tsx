import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { Map } from '../map'
import { TransitLayer } from '../transitLayer'
import { Marker } from '../marker'
import { loadGoogleMapsAPI } from '../client'
import { act } from 'react-dom/test-utils'

// tslint:disable-next-line: no-floating-promises
loadGoogleMapsAPI({
  key: 'api-key'
})

describe('Hello Map', () => {
  it('renders', async () => {
    const places = getPlaces()
    const marks = places.map((m, index) => {
      return <Marker
        key={m.id}
        icon={index === 0 ? activeIcon : inactiveIcon}
        title={'marker id: ' + m.id}
        position={{ lat: m.lat, lng: m.lng }}
      />
    })
    let mounted: ShallowWrapper<any, any, any> = undefined as any
    act(() => {
      mounted = shallow(<div>
        Google Map Test
      <Map
          zoom={10}
          center={{ lat: places[0].lat, lng: places[0].lng }}
        >
          {marks}
          <TransitLayer enabled={true} />
        </Map>
      </div>)
    })

    expect(mounted).toBeDefined()
    expect(mounted).toMatchSnapshot()
    // expect(mounted.find('.map-container').length).toBe(1)
    // expect(mounted.find('.map-ref').length).toBe(1)

    await new Promise(r => setTimeout(r, 3000))

    act(() => {
      mounted.update()
    })

    expect(mounted).toMatchSnapshot()
  })
})

export default function getPlaces () {
  const cnt = 10
  const markers = []
  const base = { lat: 35.6432027, lng: 139.6729435 }
  for (let i = 0; i < cnt; i++) {
    markers.push({
      id: i,
      title: 'marker: ' + i,
      lat: base.lat + 0.04 * i,
      lng: base.lng + 0.04 * i
    })
  }
  return markers
}

const activeIcon =
  'https://a0.muscache.com/airbnb/static/select_pdp/home_icon-9999d1852c239e9a93c7d7975441c254.png'
const inactiveIcon =
  'https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png'
