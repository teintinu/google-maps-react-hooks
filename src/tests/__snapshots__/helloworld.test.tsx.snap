// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hello Map renders 1`] = `
<div>
  Google Map Test
  <Component
    center={
      Object {
        "lat": 35.6432027,
        "lng": 139.6729435,
      }
    }
    zoom={10}
  >
    <Component
      icon="https://a0.muscache.com/airbnb/static/select_pdp/home_icon-9999d1852c239e9a93c7d7975441c254.png"
      key="0"
      position={
        Object {
          "lat": 35.6432027,
          "lng": 139.6729435,
        }
      }
      title="marker id: 0"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="1"
      position={
        Object {
          "lat": 35.6832027,
          "lng": 139.7129435,
        }
      }
      title="marker id: 1"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="2"
      position={
        Object {
          "lat": 35.7232027,
          "lng": 139.75294350000001,
        }
      }
      title="marker id: 2"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="3"
      position={
        Object {
          "lat": 35.7632027,
          "lng": 139.7929435,
        }
      }
      title="marker id: 3"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="4"
      position={
        Object {
          "lat": 35.8032027,
          "lng": 139.8329435,
        }
      }
      title="marker id: 4"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="5"
      position={
        Object {
          "lat": 35.843202700000006,
          "lng": 139.8729435,
        }
      }
      title="marker id: 5"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="6"
      position={
        Object {
          "lat": 35.883202700000005,
          "lng": 139.9129435,
        }
      }
      title="marker id: 6"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="7"
      position={
        Object {
          "lat": 35.923202700000004,
          "lng": 139.9529435,
        }
      }
      title="marker id: 7"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="8"
      position={
        Object {
          "lat": 35.963202700000004,
          "lng": 139.9929435,
        }
      }
      title="marker id: 8"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="9"
      position={
        Object {
          "lat": 36.0032027,
          "lng": 140.03294350000002,
        }
      }
      title="marker id: 9"
    />
    <TransitLayer
      enabled={true}
    />
  </Component>
</div>
`;

exports[`Hello Map renders 2`] = `
<div>
  Google Map Test
  <Component
    center={
      Object {
        "lat": 35.6432027,
        "lng": 139.6729435,
      }
    }
    zoom={10}
  >
    <Component
      icon="https://a0.muscache.com/airbnb/static/select_pdp/home_icon-9999d1852c239e9a93c7d7975441c254.png"
      key="0"
      position={
        Object {
          "lat": 35.6432027,
          "lng": 139.6729435,
        }
      }
      title="marker id: 0"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="1"
      position={
        Object {
          "lat": 35.6832027,
          "lng": 139.7129435,
        }
      }
      title="marker id: 1"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="2"
      position={
        Object {
          "lat": 35.7232027,
          "lng": 139.75294350000001,
        }
      }
      title="marker id: 2"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="3"
      position={
        Object {
          "lat": 35.7632027,
          "lng": 139.7929435,
        }
      }
      title="marker id: 3"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="4"
      position={
        Object {
          "lat": 35.8032027,
          "lng": 139.8329435,
        }
      }
      title="marker id: 4"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="5"
      position={
        Object {
          "lat": 35.843202700000006,
          "lng": 139.8729435,
        }
      }
      title="marker id: 5"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="6"
      position={
        Object {
          "lat": 35.883202700000005,
          "lng": 139.9129435,
        }
      }
      title="marker id: 6"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="7"
      position={
        Object {
          "lat": 35.923202700000004,
          "lng": 139.9529435,
        }
      }
      title="marker id: 7"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="8"
      position={
        Object {
          "lat": 35.963202700000004,
          "lng": 139.9929435,
        }
      }
      title="marker id: 8"
    />
    <Component
      icon="https://maps.gstatic.com/mapfiles/api-3/images/spotlight-poi2.png"
      key="9"
      position={
        Object {
          "lat": 36.0032027,
          "lng": 140.03294350000002,
        }
      }
      title="marker id: 9"
    />
    <TransitLayer
      enabled={true}
    />
  </Component>
</div>
`;
